linux/hid.c

```diff
@@ -942,26 +948,239 @@ void  HID_API_EXPORT hid_free_enumeration(struct hid_device_info *devs)
 	}
 }
 
+struct hid_hotplug_callback {
+    hid_hotplug_callback_handle handle;
+    unsigned short vendor_id;
+    unsigned short product_id;
+    hid_hotplug_event events;
+    void *user_data;
+    hid_hotplug_callback_fn callback;
+
+    /* Pointer to the next notification */
+    struct hid_hotplug_callback *next;
+};
+
+static struct hid_hotplug_context {
+    /* UDEV context that handles the monitor */
+    struct udev* udev_ctx;
+
+    /* UDEV monitor that receives events */
+    struct udev_monitor* mon;
+
+    /* File descriptor for the UDEV monitor that allows to check for new events with select() */
+    int monitor_fd;
+
+    /* Thread for the UDEV monitor */
+    pthread_t* thread;
+
+    pthread_mutex_t mutex;
+
+    /* HIDAPI unique callback handle counter */
+    hid_hotplug_callback_handle next_handle;
+
+    /* Linked list of the hotplug callbacks */
+    struct hid_hotplug_callback *hotplug_cbs;
+
+    /* Linked list of the device infos (mandatory when the device is disconnected) */
+    struct hid_device_info *devs;
+} hid_hotplug_context = {
+    .udev_ctx = NULL,
+    .thread = NULL,
+    .monitor_fd = -1,
+    .next_handle = 1,
+    .hotplug_cbs = NULL,
+    .devs = NULL
+};
+
+static void hid_internal_invoke_callbacks(struct hid_device_info* info, hid_hotplug_event event)
+{
+    struct hid_hotplug_callback **current = &hid_hotplug_context.hotplug_cbs;
+    while(*current) {
```

Comment:
```suggestion
    while (*current) {
```
always a space after `if`/`for`/`while`/etc.